apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -0,0 +1,228 @@
+	// ResourceFlavor. If a matching ResourceFlavor does not exist, the
+	// ClusterQueue will have an Active condition set to False.
+	Name ResourceFlavorReference `json:"name"`
+
+	// resources is the list of quotas for this flavor per resource.
+	// There could be up to 16 resources.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	Resources []ResourceQuota `json:"resources"`
+}
+
+type ResourceQuota struct {
+	// name of this resource.
+	Name corev1.ResourceName `json:"name"`
+
+	// quota is the quantity of this resource that is available for Workloads
+	// admitted by this ClusterQueue at a point in time.
+	// The quota must non-negative.
+	//
+	// If the ClusterQueue belongs to a cohort, the sum of the quotas for each
+	// (flavor, resource) combination defines the maximum quantity that can be
+	// allocated by a ClusterQueue in the cohort.
+	Quota resource.Quantity `json:"quota"`
+
+	// borrowingLimit is the maximum amount of quota for the (flavor, resource)
+	// combination that this ClusterQueue is allowed to borrow from the unused
+	// quota of other ClusterQueues in the same cohort.
+	// If null, it means that there is no borrowing limit.
+	// If not null, it must be non-negative.
+	// borrowingLimit must be null if spec.cohort is empty.
+	// +optional
+	BorrowingLimit *resource.Quantity `json:"borrowingLimit,omitempty"`
+}
+
+// ResourceFlavorReference is the name of the ResourceFlavor.
+type ResourceFlavorReference string
+
+// ClusterQueueStatus defines the observed state of ClusterQueue
+type ClusterQueueStatus struct {
+	// flavorUsage are the used quotas by flavor currently in use by the workloads
+	// assigned to this ClusterQueue.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	// +optional
+	FlavorUsage []FlavorUsage `json:"flavorUsage"`
+
+	// PendingWorkloads is the number of workloads currently waiting to be
+	// admitted to this clusterQueue.
+	// +optional
+	PendingWorkloads int32 `json:"pendingWorkloads"`
+
+	// AdmittedWorkloads is the number of workloads currently admitted to this
+	// clusterQueue and haven't finished yet.
+	// +optional
+	AdmittedWorkloads int32 `json:"admittedWorkloads"`
+
+	// conditions hold the latest available observations of the ClusterQueue
+	// current state.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type FlavorUsage struct {
+	// name of the flavor.
+	Name string `json:"name"`
+
+	// resources lists the quota usage for the resources in this flavor.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	Resources []ResourceUsage `json:"resources"`
+}
+
+type ResourceUsage struct {
+	// name of the resource
+	Name corev1.ResourceName `json:"name"`
+
+	// Total is the total quantity of the resource used, including resources
+	// borrowed from the cohort.
+	Total resource.Quantity `json:"total,omitempty"`
```

Comment:
Here seems to be a bit of inconsistency. In spec we have base + borrowed. Here is total + borrowed.

Reply:
That is a good point, but I think it's rather useful:
If an administrator considers increasing the quota, it's useful for them to directly see the total usage.
And observing the borrowed quota is useful for adjusting borrowingLimit.